core/src/main/java/org/apache/iceberg/schema/UnionByNameVisitor.java

```diff
@@ -90,6 +94,15 @@ public Boolean struct(
               Types.NestedField field = fields.get(pos);
               if (isMissing) {
                 addColumn(partnerId, field);
```

Comment:
@amogh-jahagirdar This is the core problem with the existing code.  Additional columns, no matter where they are in the schema are just added.  This always adds them to the end